src/data.cc

```diff
@@ -114,4 +114,84 @@ Status BCFData::dataset_range_and_header(const string& dataset, const range& pos
     return dataset_range(dataset, hdr.get(), pos, records);
 }
 
+// default sampleset_range implementation:
+
+// Return one iterator per 100kbp of the requested range. Each iterator simply
+// calls dataset_range_in_header on each dataset in the 100kb block.
+
+class DefaultRangeBCFIteratorImpl : public RangeBCFIterator {
+    BCFData& data_;
+    range range_;
+    bool first_range_;
+    shared_ptr<const set<string>> datasets_;
+    set<string>::const_iterator it_;
+
+public:
+    DefaultRangeBCFIteratorImpl(BCFData& data, range range, bool first_range, shared_ptr<const set<string>>& datasets)
+        : data_(data), range_(range), first_range_(first_range), datasets_(datasets), it_(datasets->begin()) {}
+
+    Status next(string& dataset, shared_ptr<const bcf_hdr_t>& hdr,
+                vector<shared_ptr<bcf1_t>>& records) override {
+        if (it_ == datasets_->end()) {
+            return Status::NotFound();
+        }
+        dataset = *it_++;
+
+        vector<shared_ptr<bcf1_t>> all_records;
+        Status s = data_.dataset_range_and_header(dataset, range_, hdr, all_records);
+        if (s.bad()) {
+             if (s == StatusCode::NOT_FOUND) {
+                // censor this error so caller doesn't think this is the normal
+                // end of iteration
+                return Status::Failure("RangeBCFIterator::next(): unexpected NotFound error", dataset);
+            }
+            return s;
+        }
+
+        // subtlety: to ensure the iterators together return each record
+        // exactly once, we may need to skip records that don't begin within
+        // range_ (unless this is the first or lowest-range iterator)
+        assert(std::is_sorted(all_records.begin(), all_records.end(),
+                              [] (shared_ptr<bcf1_t>& p1, shared_ptr<bcf1_t>& p2) {
+                                return range(p1) < range(p2);
+                              }));
+        int skip=0;
+        if (!first_range_) {
+            for (; skip<all_records.size(); skip++) {
+                range rng(all_records[skip]);
+                assert(range_.overlaps(rng));
+                if (rng.beg >= range_.beg) {
+                    break;
+                }
+            }
+        }
+        records.clear();
+        records.reserve(all_records.size()-skip);
```

Comment:
Sorry, I don't understand what this line is trying to achieve. Perhaps you can light here?